backends/bmv2/portable_switch/portableSwitch.cpp

```diff
@@ -0,0 +1,823 @@
+        else if (info.first == INGRESS && info.second == DEPARSER)
+            pinfo->deparsers.emplace("ingress", c);
+        else if (info.first == EGRESS && info.second == DEPARSER)
+            pinfo->deparsers.emplace("egress", c);
+    }
+}
+
+void PortableSwitchBackend::convert(const IR::ToplevelBlock* tlb) {
+    CHECK_NULL(tlb);
+    PsaProgramStructure structure(refMap, typeMap);
+
+    auto parsePsaArch = new ParsePsaArchitecture(&structure);
+    auto main = tlb->getMain();
+    if (!main) return;
+    main->apply(*parsePsaArch);
+
+    auto evaluator = new P4::EvaluatorPass(refMap, typeMap);
+    auto program = tlb->getProgram();
+    PassManager simplify = {
+        /* TODO */
+        // new RenameUserMetadata(refMap, userMetaType, userMetaName),
+        new P4::ClearTypeMap(typeMap),  // because the user metadata type has changed
+        // new P4::SynthesizeActions(refMap, typeMap, new SkipControls(&non_pipeline_controls)),
+        new P4::MoveActionsToTables(refMap, typeMap),
+        new P4::TypeChecking(refMap, typeMap),
+        new P4::SimplifyControlFlow(refMap, typeMap),
+        new LowerExpressions(typeMap),
+        new P4::ConstantFolding(refMap, typeMap, false),
+        new P4::TypeChecking(refMap, typeMap),
+        // new RemoveComplexExpressions(refMap, typeMap, new ProcessControls(&pipeline_controls)),
+        new P4::SimplifyControlFlow(refMap, typeMap),
+        new P4::RemoveAllUnusedDeclarations(refMap),
+        evaluator,
+        new VisitFunctor([this, evaluator, structure]() {
+            toplevel = evaluator->getToplevelBlock(); }),
+    };
+    program->apply(simplify);
+
+    // map IR node to compile-time allocated resource blocks.
+    toplevel->apply(*new BMV2::BuildResourceMap(&structure.resourceMap));
+
+    PassManager toJson = {
+        new DiscoverStructure(&structure),
+        new InspectPsaProgram(refMap, typeMap, &structure),
+        new ConvertPsaToJson(refMap, typeMap, toplevel, json, &structure)
+    };
+    program->apply(toJson);
+
+    json->add_program_info(options.file);
+    json->add_meta_info();
+}
+
+EXTERN_CONVERTER_SINGLETON(Hash)
+EXTERN_CONVERTER_SINGLETON(Checksum)
+EXTERN_CONVERTER_SINGLETON(InternetChecksum)
+EXTERN_CONVERTER_SINGLETON(Counter)
+EXTERN_CONVERTER_SINGLETON(DirectCounter)
+EXTERN_CONVERTER_SINGLETON(Meter)
+EXTERN_CONVERTER_SINGLETON(DirectMeter)
+EXTERN_CONVERTER_SINGLETON(Register)
+EXTERN_CONVERTER_SINGLETON(Random)
+EXTERN_CONVERTER_SINGLETON(ActionProfile)
+EXTERN_CONVERTER_SINGLETON(ActionSelector)
+EXTERN_CONVERTER_SINGLETON(Digest)
+
+CONVERT_EXTERN_OBJECT(Hash) {
+    auto primitive = mkPrimitive("Hash");
+    return primitive;
+}
+
+CONVERT_EXTERN_OBJECT(Checksum) {
+    auto primitive = mkPrimitive("Checksum");
+    return primitive;
+}
+
+CONVERT_EXTERN_OBJECT(InternetChecksum) {
+    auto primitive = mkPrimitive("InternetChecksum");
+    return primitive;
+}
+
+CONVERT_EXTERN_OBJECT(Counter) {
```

Comment:
This is what I am talking about: it is not clear at all what is available in the body of this function.